### DISCUSSION TITLE
Schema review!

### DISCUSSION THREAD

**Discussion by mzalevski** [colinhacks/zod#334] [General]
Hi, I thought that it would be helpful if we could review complicated schemas here. Let me start. Consider the schema below. How could I do better?

Things to think about:
1. Numeric strings validation.
2. Extracting refinement functions, e.g. `z.object({ ... }).refine(adminSecretLogic).refine(adminCodeLogic)`.
3. Overall code quality & best practices.

```typescript
import * as z from "zod";
import isNumeric from "validator/lib/isNumeric";

interface User {
  id: string;
  isAdmin: boolean;
  adminSecret?: string;
  adminCode?: string;
}

const userSchema: z.ZodSchema<User> = z
  .object({
    id: z.string(),
    isAdmin: z.boolean(),
    adminSecret: z.string().max(20).optional(),
    adminCode: z.string().length(4).optional(),
  })
  .refine(data => (data.isAdmin ? data.adminSecret : true), {
    message: "Admin secret is missing from an admin user.",
    path: ["adminSecret"],
  })
  .refine(
    data =>
      data.adminSecret && data.adminSecret.length > 10
        ? data.adminCode && isNumeric(data.adminCode)
        : true,
    data => ({
      message:
        data.adminCode && !isNumeric(data.adminCode)
          ? "Admin code needs to be digits only."
          : "4-digit admin code is missing from an admin user with a long secret.",
      path: ["adminCode"],
    })
  );

// Admin secret is missing from an admin user.
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: true,
});

// 4-digit admin code is missing from an admin user with a long secret.
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: true,
  adminSecret: "moreThenTenChars",
});

// Admin code needs to be digits only.
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: true,
  adminSecret: "moreThenTenChars",
  adminCode: "a134",
});

// No errors!
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: false,
});

// No errors!
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: true,
  adminSecret: "belowTen",
});

// No errors!
userSchema.parse({
  id: "20ec0677-68d8-45b5-a522-16cff767ac6f",
  isAdmin: true,
  adminSecret: "moreThenTenChars",
  adminCode: "1234",
});
```

**Comment by prescience-data** [comment]
It would make the code easier to read with a ```typescript prefix on the code block.

Is there a reason you are manually defining `User` rather than inferring it?

```typescript

export const UserSchema = z
  .object({
    id: z.string(),
    isAdmin: z.boolean().default(false),
    adminSecret: z.string().max(20).optional(),
    adminCode: z.string().length(4).optional(),
  })

export type User = z.infer<typeof UserSchema>


const myUser: User = UserSchema.parse({ id: "foo" })
```

**Reply by mzalevski** [reply]
I tend to manually define interfaces, because a parsed object is recognized as e.g. `User` without an additional declaration like `const myUser: User = ...`.

I've added ```typescript. Thanks!

**Reply by prescience-data** [reply]
This was the line I meant though:
```typescript
export type User = z.infer<typeof UserSchema>
```
===
```typescript
export interface User {
  id: string;
  isAdmin: boolean;
  adminSecret?: string;
  adminCode?: string;
}
```
That is effectively identical to your manually described `User` interface but without the duplication.

It also ensures a tight 1:1 binding of `schema <> type interface`, which eliminates the risk of accidental mismatches if you update the schema but forget to update the type  (or vice versa).

**Reply by mzalevski** [reply]
I know that you meant the above. I mean it too :)
I know that these are equivalent but:
1. if you infer the `User` interface, then after you `userSchema.parse(obj)`, the returned object is not recognized as `User` but as an object literal with all the fields from interface (**not as the interface tough!**)
2. using `const userSchema = ZodSchema<User> = object({ ... })` declaration also ensures a tight 1:1 binding. It won't compile if the `userSchema` is not compliant with the `User` interface.

**Reply by prescience-data** [reply]
Ah, interesting, so this actually doesn't `===`? More like a loose `==`?

```typescript
export type User = z.infer<typeof UserSchema>
```
===
```typescript
export interface User {
  id: string;
  isAdmin: boolean;
  adminSecret?: string;
  adminCode?: string;
}
```

**Reply by mzalevski** [reply]
I would not say it's loose. It is indeed tightly coupled. I think that what I am talking about is simply due to order of operations. The `User` interface does not yet exist at the moment of defining the schema. Hence the schema has no way to know that you would really like to see `User` in your IDE intelisense :)

**Reply by prescience-data** [reply]
Ah, so it's more a preference thing? 
I always define my schemas first as concrete, then export the inferred interfaces (and in turn use those throughout the application).
It sounds like the distinction is that you prefer the reverse where the schema is subordinate to an existing interface / type?

**Reply by mzalevski** [reply]
Correct :)
